packages/mobile/src/identity/feelessVerification.ts

```diff
@@ -0,0 +1,1061 @@
+
+  yield put(refreshAllBalances())
+}
+
+export function* feelessRestartableVerification(initialWithoutRevealing: boolean) {
+  let isRestarted = false
+  while (true) {
+    const withoutRevealing = !isRestarted && initialWithoutRevealing
+    yield call(navigate, Screens.VerificationLoadingScreen, {
+      withoutRevealing,
+    })
+    yield put(feelessResetVerification())
+    yield call(getConnectedAccount)
+    if (isRestarted || (yield select(isFeelessVerificationStateExpiredSelector))) {
+      yield call(feelessFetchVerificationState)
+    }
+
+    const { verification, restart } = yield race({
+      verification: call(feelessDoVerificationFlow, withoutRevealing),
+      restart: take(Actions.FEELESS_RESEND_ATTESTATIONS),
+    })
+
+    if (restart) {
+      isRestarted = true
+      const { status }: FeelessVerificationState = yield select(feelessVerificationStateSelector)
+      ValoraAnalytics.track(VerificationEvents.verification_resend_messages, {
+        count: status.numAttestationsRemaining,
+      })
+    } else {
+      return verification
+    }
+    Logger.debug(TAG, 'Verification has been restarted')
+  }
+}
+
+export function* feelessDoVerificationFlow(withoutRevealing: boolean = false) {
+  Logger.debug(TAG, '@feelessDoVerificationFlow', 'Starting feeless verification')
+  let receiveMessageTask: Task | undefined
+  let autoRetrievalTask: Task | undefined
+
+  try {
+    const [walletAddress, contractKit, e164Number]: [string, ContractKit, string] = yield all([
+      call(getConnectedUnlockedAccount),
+      call(getContractKit),
+      select(e164NumberSelector),
+      put(feelessSetVerificationStatus(VerificationStatus.Prepping)),
+    ])
+
+    let feelessVerificationState: FeelessVerificationState = yield select(
+      feelessVerificationStateSelector
+    )
+
+    const komenciKit = new KomenciKit(contractKit, walletAddress, {
+      url: KOMENCI_URL,
+      token: feelessVerificationState.komenci.sessionToken,
+    })
+
+    // Start by checking again to make sure Komenci is ready. Throws error if not
+    yield call(fetchKomenciReadiness, komenciKit)
+
+    // There should be no instances where the first param is truethy but the second isn't.
+    // Mainly including the second param to satisfy typescript
+    if (
+      !feelessVerificationState.status.isVerified ||
+      !feelessVerificationState.komenci.unverifiedMtwAddress
+    ) {
+      // Start or resume a Komenci session and update state. Throws error if unable to do so
+      yield call(startOrResumeKomenciSession, komenciKit, e164Number)
+
+      // Adds phoneHash into verification state, will ping Komenci for it only if needed.
+      // Throws an error if unable to get phoneHash
+      yield call(fetchPhoneHashDetails, komenciKit, e164Number)
+
+      const startingPepperQuota = feelessVerificationState.komenci.pepperQuotaRemaining
+      // Now that we are guarnateed to have the phoneHash, check again to see if the
```

Comment:
This feels like it should belong in `feelessFetchVerificationState`...

Reply:
Are you suggesting we move the pepper query into `feelessVerificationState`? I agree that that feels like a better home for it from a "separation of concerns" perspective but practically speaking I think it makes less sense given how these functions are currently tied into the UI.

The user is waiting for `feelessFetchVerificationState` to finish before they can even press "start" on the "verification start" screen. For that reason I am reluctant to add a non-state related Komenci call that will further slow things down. While it may look expensive to be repeating function calls, remember that many of these will quickly bail out if the right conditions aren't being met.